crates/sui-data-ingestion-core/src/reducer.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Reducer, Worker};
+use anyhow::Result;
+use std::collections::HashMap;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::mpsc;
+
+pub(crate) async fn reduce<W: Worker>(
+    task_name: String,
+    mut current_checkpoint_number: CheckpointSequenceNumber,
+    mut progress_receiver: mpsc::Receiver<(CheckpointSequenceNumber, W::Result)>,
+    executor_progress_sender: mpsc::Sender<(String, CheckpointSequenceNumber)>,
+    reducer: Option<Box<dyn Reducer<W::Result>>>,
+) -> Result<()> {
+    let mut unprocessed = HashMap::new();
+    let mut batch = vec![];
+    let mut progress_update = None;
+
+    while let Some((checkpoint_number, message)) = progress_receiver.recv().await {
```

Comment:
if there is no new data from recv, unprocessed may starve, which is not common in prod but very common in testing, let's do now_or_never？ https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never

Reply:
+1. We should use a `loop` as the outside loop. When receiving, we should try to receive for as much as we can until there is no more in the channel.